### ISSUE TITLE
queries that use AsExpandable and .ToListAsync() do NOT run asynchronously

### ISSUE THREAD

**Issue by lvanderbijl** [scottksmith95/LINQKit#7]
I'm running a number of queries that take advantage of .invoke to re-use where clauses as well as select projections.

However, I've found that the ExpandableDBAsyncEnumerator doesn't actually run the queries asynchronously. It forces the expression to be run synchronously. 

This is a pretty major issue for me. Has anyone else come across this? Does anyone have a suitable fix?

ie.

public async Task<object> GetObjectAsync()
{
      var obj = await dbSet.AsExpandable().FirstOrDefault();
return obj;

}

var t1 = GetObjectAsync();
var t2 = GetObjectAsync();

await Task.WhenAll(t1,t2);

These two calls to getobjectasync will NOT run asynchrounously - they'll only run synchronously.

Any ideas?

**Comment by Thorium** [comment]
Beware that there is EntityFramwork limitation:
http://stackoverflow.com/questions/24702183/does-entity-framework-support-parallel-async-queries

        [Fact]
        public async Task ExpressionAsyncTest()
        {
            Expression<Func<int, bool>> condition = a => a % 2 == 0;

            var expanded = condition.Expand();
            
            var dbset = new[] {1, 2, 3, 4}.AsQueryable(); // from db
            var xs = dbset.AsExpandable();

            var result =
                from x in xs
                where expanded.Invoke(x)
                select x;

            var t1 = result.ToListAsync();
            var t2 = result.Skip(1).ToListAsync();

            Assert.Equal(2, (await t1).First());
            Assert.Equal(4, (await t2).First());
        }

As far as I know, LINQKit is not the one that limits you...

**Comment by axelheer** [comment]
Running queries in parallel may not be supported, but running a single query async still has a lot of benefits since the current thread becomes free in the meantime; thus, he's available for your application or your host (web server, ui thread, ...).

The problem with the current implementation is, that the async enumerator / provider of *Entity Framework* isn't called, a synchronous `Task.FromResult` is used instead. A detection if the current enumerator / provider is in fact async and an appropriate direct call should solve this.

The pull request i've submitted will help.

**Comment by lvanderbijl** [comment]
Thanks for the responses.

@vinneyk Yeah, you're right I wrote the code down incorrectly - it should have read FirstOrDefaultAsync.

@Thorium You're right, I can't make 2 calls to the same dbset at the same time - I realised that after I lodged this issue. I ended up solving that by using multiple contexts.

However, as @axelheer notes, there's still an issue actually running the code (even a single call) asynchronously.

If/when I get a chance I'll look at the pull request as that would be extremely helpful. Because of this issue I basically had to remove all my linqkit code entirely. It would be nice to be able to use it again.

Thanks Axel!
